src/lib.rs

```diff
@@ -287,69 +327,518 @@ impl<D: AsMut<[Digit]>> BigInt<D> {
+        debug_assert_eq!(scratch_space_size, ScratchSpace::division(n, m));
+
+        // Allocate space on the heap, split it into `output` that stands for addition result and
+        // `scratch_space` that is auxiliary buffer for gmp functions
+        let mut buffer = vec![Digit::zero(); output_size + scratch_space_size];
+        let (output, scratch_space) = buffer.split_at_mut(output_size);
+        let mut output = BigInt::from_digits(output);
+        let scratch_space = ScratchSpace::new(scratch_space);
+
+        let mut a_bigint = BigInt::from_digits(a);
+        unsafe {
+            // Safety:
+            // * We checked that n>=m and that n and m are non-zero (see match statement above)
+            // * We allocated sufficient memory for both output and scratch space
+            output.div_unchecked(&mut a_bigint, &BigInt::from_digits(&b[..m]), scratch_space);
+        };
+
+        // Note: we are comparing two size_t integers here, we assume it's constant-time
+        let last_digit_is_zero = Choice::from(u8::from(output.as_ref()[n - m] == Digit::zero()));
+        let output_size = size_t::conditional_select(
+            &((n - m + 1) as size_t),
+            &((n - m) as size_t),
+            last_digit_is_zero,
+        ) as usize;
+
+        // Now we need to resize `buffer` to `output_size`,
+        // ie. we want to strip the allocated scratch space.
+        buffer[output_size..].zeroize();
+        buffer.resize(output_size, Digit::zero());
+        // Also zeroize `a` that has been allocated earlier
+        a_bigint.as_mut().zeroize();
+
+        BigInt::from_digits(buffer.into_boxed_slice())
+    }
+}
+
+impl<D: AsMut<[Digit]>> BigInt<D> {
+    /// Computes `floor(a / b)`, writes the result to `self` and remainder `a mod b` - to
+    /// the `len(b)` most significant limbs of `a`. Accordingly, `a` must be mutable.
+    ///
+    /// It is a low-level function allowing you to manage allocations. Its safe analogue is
+    /// [`&a / &b`](#impl-Div%3C%26%27_%20BigInt%3CD2%3E%3E).
+    ///
+    /// ## Safety
+    /// Assuming that `a` and `b` have [sizes][size] `n` and `m` respectively:
+    /// * `n >= m > 0`
+    /// * [Size][size] of `self` must be exactly `n-m` limbs
+    /// * Size of `scratch_space` must be exactly
+    /// [`ScratchSpace::division(n, m)`](ScratchSpace::division) limbs
+    ///
+    /// Note: we check that these requirements are met using debug assertions.
+    ///
+    /// [size]: BigInt::size
+    ///
+    /// ## Constant time
+    /// This function is constant time in size of its arguments, ie. number of instructions and
+    /// memory access patterns remain the same for the same `n` and `m`.
+    pub unsafe fn div_unchecked(
+        &mut self,
+        a: &mut BigInt<impl AsMut<[Digit]>>,
+        b: &BigInt<impl AsRef<[Digit]>>,
+        mut scratch_space: ScratchSpace,
+    ) {
+        let n = a.as_mut().len() - b.size();
```

Comment:
Heck, I just realised that we cannot use methods of `BigInt<D: AsRef<[Digit]>>` on `a`. I'll try to figure out how that can be fixed. Let me know if you have any ideas 🙂

Reply:
If I understand you correctly and you're talking about the fact that `gmp` mutates `a` in division (writes remainder into its most significant limbs to be precise), then I guess we can only clone `a`, there are not many other options? I might be wrong though.

Reply:
Sorry for confusion, I mean the `.size()` method. Basically, you have `a` and `b` both `BigInt`s, and `.size()` is defined for `BigInt`. But here you can't write `a.size()`, you have to write `a.as_refe().len()`, which is disappointing.

Reply:
Ah, yes. A bit unpleasant, but probably not critical?

Reply:
Not critical